File: bindings/r/meshioplusplus/src/sequence.c
/* Multi-file / transient datasets.
 *
 * A sequence is an ordered PLAN over a set of files (or the steps inside one
 * multi-step file), read one step at a time. It owns no mesh, deliberately:
 * caching what it handed out is exactly the accumulation the streaming
 * guarantee forbids, which is why `mio_sequence_read()` returns an OWNED mesh.
 *
 * Its own tag and finalizer, exactly as `mio_mesh` and `mio_xdmf_series` have,
 * so the three can never be passed for one another. A foreign or
 * already-released pointer is an R error, never a dereference. */

#include "mio_r.h"

SEXP mio_r_sequence_tag = NULL;

static void mio_r_sequence_finalizer(SEXP x) {
    mio_sequence *seq = (mio_sequence *)R_ExternalPtrAddr(x);
    if (seq != NULL) {
        mio_sequence_free(seq);
        R_ClearExternalPtr(x);
    }
}

static SEXP wrap_sequence(mio_sequence *seq) {
    SEXP ptr = PROTECT(R_MakeExternalPtr(seq, mio_r_sequence_tag, R_NilValue));
    R_RegisterCFinalizerEx(ptr, mio_r_sequence_finalizer, TRUE);
    SEXP cls = PROTECT(Rf_mkString("mio_sequence"));
    Rf_setAttrib(ptr, R_ClassSymbol, cls);
    UNPROTECT(2); /* cls, ptr */
    return ptr;
}

static mio_sequence *sequence_of(SEXP x) {
    if (TYPEOF(x) != EXTPTRSXP || R_ExternalPtrTag(x) != mio_r_sequence_tag) {
        Rf_error("expected a mio_sequence object");
    }
    mio_sequence *seq = (mio_sequence *)R_ExternalPtrAddr(x);
    if (seq == NULL) {
        Rf_error("this mio_sequence has already been released");
    }
    return seq;
}

/* "auto"/"file"/"filename"/"index" -> the C enum. */
static int32_t sequence_time_from(SEXP time_from) {
    const char *s = mio_r_string(time_from, "time_from");
    if (strcmp(s, "auto") == 0)
        return 0;
    if (strcmp(s, "file") == 0)
        return 1;
    if (strcmp(s, "filename") == 0)
        return 2;
    if (strcmp(s, "index") == 0)
        return 3;
    Rf_error("time_from must be \"auto\", \"file\", \"filename\" or \"index\"");
    return 0; /* unreachable */
}

SEXP R_mio_sequence_open(SEXP pattern, SEXP format, SEXP times, SEXP time_from, SEXP sort) {
    mio_sequence_opts opts;
    mio_sequence_opts_init(&opts);
    opts.format = mio_r_opt_string(format);
    opts.time_from = sequence_time_from(time_from);
    opts.sort = (int32_t)mio_r_bool(sort, "sort");

    double *tbuf = NULL;
    R_xlen_t ntimes = 0;
    if (times != R_NilValue) {
        SEXP nums = PROTECT(Rf_coerceVector(times, REALSXP));
        ntimes = Rf_xlength(nums);
        tbuf = REAL(nums);
        opts.times = tbuf;
        opts.num_times = (int64_t)ntimes;
        mio_sequence *seq = mio_sequence_open_ex(mio_r_string(pattern, "pattern"), &opts);
        UNPROTECT(1); /* nums */
        if (seq == NULL)
            mio_r_fail("sequence_open");
        return wrap_sequence(seq);
    }
    mio_sequence *seq = mio_sequence_open_ex(mio_r_string(pattern, "pattern"), &opts);
    if (seq == NULL)
        mio_r_fail("sequence_open");
    return wrap_sequence(seq);
}

SEXP R_mio_sequence_open_list(SEXP paths, SEXP format, SEXP times, SEXP time_from, SEXP sort) {
    if (TYPEOF(paths) != STRSXP || Rf_xlength(paths) == 0) {
        Rf_error("paths must be a non-empty character vector");
    }
    R_xlen_t n = Rf_xlength(paths);
    const char **cpaths = (const char **)R_alloc((size_t)n, sizeof(const char *));
    for (R_xlen_t i = 0; i < n; ++i) {
        cpaths[i] = CHAR(STRING_ELT(paths, i));
    }

    mio_sequence_opts opts;
    mio_sequence_opts_init(&opts);
    opts.format = mio_r_opt_string(format);
    opts.time_from = sequence_time_from(time_from);
    opts.sort = (int32_t)mio_r_bool(sort, "sort");

    if (times != R_NilValue) {
        SEXP nums = PROTECT(Rf_coerceVector(times, REALSXP));
        opts.times = REAL(nums);
        opts.num_times = (int64_t)Rf_xlength(nums);
        mio_sequence *seq = mio_sequence_open_list(cpaths, (int64_t)n, &opts);
        UNPROTECT(1); /* nums */
        if (seq == NULL)
            mio_r_fail("sequence_open_list");
        return wrap_sequence(seq);
    }
    mio_sequence *seq = mio_sequence_open_list(cpaths, (int64_t)n, &opts);
    if (seq == NULL)
        mio_r_fail("sequence_open_list");
    return wrap_sequence(seq);
}

SEXP R_mio_sequence_count(SEXP seq) {
    int64_t n = mio_sequence_count(sequence_of(seq));
    if (n < 0)
        mio_r_fail("sequence_count");
    return Rf_ScalarReal((double)n);
}

SEXP R_mio_sequence_path(SEXP seq, SEXP index) {
    mio_sequence *s = sequence_of(seq);
    int64_t i = (int64_t)mio_r_int(index, "index") - 1; /* R is 1-based */
    int64_t need = mio_sequence_path(s, i, NULL, 0);
    if (need < 0)
        mio_r_fail("sequence_path");
    char *buf = (char *)R_alloc((size_t)need + 1, 1);
    if (mio_sequence_path(s, i, buf, need + 1) < 0)
        mio_r_fail("sequence_path");
    return Rf_mkString(buf);
}

SEXP R_mio_sequence_step(SEXP seq, SEXP index) {
    int64_t k = mio_sequence_step(sequence_of(seq), (int64_t)mio_r_int(index, "index") - 1);
    if (k < 0)
        mio_r_fail("sequence_step");
    return Rf_ScalarReal((double)k);
}

SEXP R_mio_sequence_time(SEXP seq, SEXP index) {
    double t = 0.0;
    mio_r_check(mio_sequence_time(sequence_of(seq), (int64_t)mio_r_int(index, "index") - 1, &t),
                "sequence_time");
    return Rf_ScalarReal(t);
}

SEXP R_mio_sequence_time_source(SEXP seq, SEXP index) {
    int32_t code =
        mio_sequence_time_source(sequence_of(seq), (int64_t)mio_r_int(index, "index") - 1);
    if (code < 0)
        mio_r_fail("sequence_time_source");
    static const char *names[] = {"explicit", "file", "filename", "index"};
    return Rf_mkString(names[code]);
}

SEXP R_mio_sequence_read(SEXP seq, SEXP index) {
    mio_mesh *m = mio_sequence_read(sequence_of(seq), (int64_t)mio_r_int(index, "index") - 1);
    if (m == NULL)
        mio_r_fail("sequence_read");
    return mio_r_wrap_mesh(m);
}

SEXP R_mio_sequence_free(SEXP seq) {
    if (TYPEOF(seq) == EXTPTRSXP && R_ExternalPtrTag(seq) == mio_r_sequence_tag) {
        mio_sequence *s = (mio_sequence *)R_ExternalPtrAddr(seq);
        if (s != NULL) {
            mio_sequence_free(s);
            R_ClearExternalPtr(seq);
        }
    }
    return R_NilValue;
}

SEXP R_mio_sequence_to_timeseries(SEXP seq, SEXP out_path, SEXP out_format, SEXP ascii) {
    /* The transient writer drives XdmfTimeSeriesWriter directly rather than
     * the registry, so `ascii` is the one write option with anywhere to go:
     * it selects XDMF's "XML" data format (no HDF5 needed) over the default
     * "HDF" -- the option a build without HDF5 support (this package's own
     * notebook environment among them, see doc/r.md) needs. */
    if (!mio_r_bool(ascii, "ascii")) {
        mio_r_check(mio_sequence_to_timeseries(sequence_of(seq), mio_r_string(out_path, "out_path"),
                                               mio_r_opt_string(out_format)),
                    "sequence_to_timeseries");
        return R_NilValue;
    }
    mio_write_opts opts;
    mio_write_opts_init(&opts);
    opts.encoding = MIO_ENCODING_ASCII;
    mio_r_check(mio_sequence_to_timeseries_ex(sequence_of(seq), mio_r_string(out_path, "out_path"),
                                              mio_r_opt_string(out_format), &opts),
                "sequence_to_timeseries");
    return R_NilValue;
}

SEXP R_mio_timeseries_to_sequence(SEXP in_path, SEXP in_format, SEXP out_pattern, SEXP out_format) {
    mio_r_check(mio_timeseries_to_sequence(
                    mio_r_string(in_path, "in_path"), mio_r_opt_string(in_format),
                    mio_r_string(out_pattern, "out_pattern"), mio_r_opt_string(out_format)),
                "timeseries_to_sequence");
    return R_NilValue;
}

SEXP R_mio_sequence_pipeline_run_file(SEXP settings_path) {
    mio_r_check(mio_sequence_pipeline_run_file(mio_r_string(settings_path, "settings_path")),
                "sequence_pipeline_run_file");
    return R_NilValue;
}

SEXP R_mio_sequence_pipeline_run_json(SEXP json_text) {
    mio_r_check(mio_sequence_pipeline_run_json(mio_r_string(json_text, "json_text")),
                "sequence_pipeline_run_json");
    return R_NilValue;
}

File: bindings/julia/MeshioPlusPlus/src/sequence.jl
# Multi-file / transient datasets: an ordered plan over a set of files (or the
# steps inside one multi-step file), read one step at a time.
#
# See doc/sequences.md. The C-side handle stores paths, per-file step indices
# and time values, and never a mesh -- which is why `read_step` returns an
# owning `Mesh` rather than a borrow the sequence would have to cache.

"""
    Sequence(pattern; format="", times=nothing, time_from="auto")
    Sequence(paths::AbstractVector{<:AbstractString}; ..., sort=false)

An ordered, lazily-read multi-file dataset.

Built from a glob `pattern` (`*` and `?` only -- **no** `**` or `[set]`; the
directory part is taken literally) or from an explicit list of `paths`, whose
order is the caller's and is kept unless `sort=true`.

Ordering is **natural-numeric**, so `out_9.vtu` precedes `out_10.vtu` — a
lexicographic sort gets that backwards, which is why the rule is documented
rather than left to the filesystem.

Each entry carries a time value and a record of where it came from
([`time_source`](@ref)): `:explicit`, `:file`, `:filename` or `:index` (the
fallback). "The file said 0.25" and "nothing said anything, so this is
position 3" are different facts.

Nothing is read until [`read_step`](@ref), and the sequence caches nothing, so
a 500-step dataset is traversable without materialising it:

```julia
seq = Sequence("out_*.vtu")
for i in 1:length(seq)
    mesh = read_step(seq, i)
    # ... one mesh alive at a time ...
end
close(seq)
```

Like [`Mesh`](@ref), the handle is released by a finalizer; [`close`](@ref)
does it deterministically and is idempotent. The `do`-block form closes it even
if the body throws.
"""
mutable struct Sequence
    ptr::Ptr{Cvoid}

    function Sequence(ptr::Ptr{Cvoid})
        s = new(ptr)
        finalizer(s) do x
            if x.ptr != C_NULL
                ccall(_sym(:mio_sequence_free), Cvoid, (Ptr{Cvoid},), x.ptr)
                x.ptr = C_NULL
            end
        end
        s
    end
end

# Mirror of C `mio_sequence_opts`; field order and types are ABI.
struct _SequenceOpts
    format::Cstring
    times::Ptr{Cdouble}
    num_times::Int64
    time_from::Int32
    sort::Int32
    reserved::NTuple{6,Int64}
end

const _TIME_FROM = Dict("auto" => Int32(0), "file" => Int32(1),
                        "filename" => Int32(2), "index" => Int32(3))
const _TIME_SOURCE = (:explicit, :file, :filename, :index)

function _sequence_opts(format, times, time_from, sort)
    code = get(_TIME_FROM, String(time_from), nothing)
    code === nothing && throw(ArgumentError(
        "time_from must be \"auto\", \"file\", \"filename\" or \"index\", got $(time_from)"))
    tvec = times === nothing ? Cdouble[] : Cdouble[Float64(t) for t in times]
    (_SequenceOpts(Base.unsafe_convert(Cstring, Base.cconvert(Cstring, String(format))),
                   isempty(tvec) ? Ptr{Cdouble}(C_NULL) : pointer(tvec),
                   Int64(length(tvec)), code, Int32(sort), ntuple(_ -> Int64(0), 6)),
     tvec)
end

function Sequence(pattern::AbstractString; format::AbstractString="",
                  times=nothing, time_from="auto")
    fmt = Base.cconvert(Cstring, String(format))
    opts, tvec = _sequence_opts(format, times, time_from, false)
    GC.@preserve fmt tvec begin
        Sequence(_check_ptr(ccall(_sym(:mio_sequence_open_ex), Ptr{Cvoid},
                                  (Cstring, Ref{_SequenceOpts}), pattern, Ref(opts))))
    end
end

function Sequence(paths::AbstractVector{<:AbstractString}; format::AbstractString="",
                  times=nothing, time_from="auto", sort::Bool=false)
    isempty(paths) && throw(ArgumentError("the path list is empty"))
    fmt = Base.cconvert(Cstring, String(format))
    opts, tvec = _sequence_opts(format, times, time_from, sort)
    cpaths = [Base.cconvert(Cstring, String(p)) for p in paths]
    ptrs = [Base.unsafe_convert(Cstring, p) for p in cpaths]
    GC.@preserve fmt tvec cpaths begin
        Sequence(_check_ptr(ccall(_sym(:mio_sequence_open_list), Ptr{Cvoid},
                                  (Ptr{Cstring}, Int64, Ref{_SequenceOpts}),
                                  ptrs, Int64(length(ptrs)), Ref(opts))))
    end
end

function Sequence(f::Function, args...; kwargs...)
    seq = Sequence(args...; kwargs...)
    try
        f(seq)
    finally
        close(seq)
    end
end

function _handle(s::Sequence)
    s.ptr == C_NULL && throw(MeshioError(MIO_ERR_INVALID_ARG, "sequence has been closed"))
    s.ptr
end

"""
    close(seq::Sequence)

Release the sequence handle deterministically. Idempotent; the finalizer does
the same thing later if you do not.
"""
function Base.close(s::Sequence)
    if s.ptr != C_NULL
        ccall(_sym(:mio_sequence_free), Cvoid, (Ptr{Cvoid},), s.ptr)
        s.ptr = C_NULL
    end
    nothing
end

"""
    length(seq::Sequence)

The number of steps.
"""
Base.length(s::Sequence) =
    Int(_check_count(ccall(_sym(:mio_sequence_count), Int64, (Ptr{Cvoid},), _handle(s)),
                     "mio_sequence_count"))

"""
    path(seq, i)

Entry `i`'s file path (1-based).
"""
function path(s::Sequence, i::Integer)
    handle = _handle(s)
    _getstring() do buf, len
        ccall(_sym(:mio_sequence_path), Int64, (Ptr{Cvoid}, Int64, Ptr{UInt8}, Int64),
              handle, Int64(i - 1), buf, len)
    end
end

"""
    step(seq, i)

Entry `i`'s step index **within its own file** — 0 for a single-step file, so a
directory of `.vtu`s and one multi-step `.xdmf` have the same shape.
"""
step(s::Sequence, i::Integer) =
    Int(_check_count(ccall(_sym(:mio_sequence_step), Int64, (Ptr{Cvoid}, Int64),
                           _handle(s), Int64(i - 1)), "mio_sequence_step"))

"""
    time(seq, i)

Entry `i`'s time value.
"""
function time(s::Sequence, i::Integer)
    out = Ref{Cdouble}(0.0)
    _check(ccall(_sym(:mio_sequence_time), Cint, (Ptr{Cvoid}, Int64, Ref{Cdouble}),
                 _handle(s), Int64(i - 1), out))
    out[]
end

"""
    time_source(seq, i)

Where entry `i`'s time came from: `:explicit`, `:file`, `:filename` or
`:index`.
"""
function time_source(s::Sequence, i::Integer)
    code = ccall(_sym(:mio_sequence_time_source), Int32, (Ptr{Cvoid}, Int64),
                 _handle(s), Int64(i - 1))
    _check_count(code, "mio_sequence_time_source")
    _TIME_SOURCE[code + 1]
end

"""
    read_step(seq, i)

Read entry `i` (1-based). The returned [`Mesh`](@ref) is **owned** by the
caller and independent of the sequence, which caches nothing — that is what
keeps a long dataset traversable.
"""
read_step(s::Sequence, i::Integer) =
    Mesh(_check_ptr(ccall(_sym(:mio_sequence_read), Ptr{Cvoid}, (Ptr{Cvoid}, Int64),
                          _handle(s), Int64(i - 1))))

# Iteration: `for mesh in seq` yields one owned mesh per step, lazily.
Base.iterate(s::Sequence, i::Int=1) = i > length(s) ? nothing : (read_step(s, i), i + 1)
Base.eltype(::Type{Sequence}) = Mesh
Base.IteratorSize(::Type{Sequence}) = Base.HasLength()

"""
    to_timeseries(seq, out_path; format="")

Fan-in: write every step into one multi-step file. Streams — one mesh alive at
a time. A format that cannot hold a series throws naming itself and pointing at
`{step}`, never a silent truncation to the first step.
"""
function to_timeseries(s::Sequence, out_path::AbstractString; format::AbstractString="")
    _check(ccall(_sym(:mio_sequence_to_timeseries), Cint, (Ptr{Cvoid}, Cstring, Cstring),
                 _handle(s), out_path, format))
    nothing
end

"""
    timeseries_to_sequence(in_path, out_pattern; in_format="", out_format="")

Fan-out: write each step of the multi-step file `in_path` to `out_pattern`,
which must contain `{step}` or `{index}`. Streams likewise.
"""
function timeseries_to_sequence(in_path::AbstractString, out_pattern::AbstractString;
                                in_format::AbstractString="", out_format::AbstractString="")
    _check(ccall(_sym(:mio_timeseries_to_sequence), Cint,
                 (Cstring, Cstring, Cstring, Cstring),
                 in_path, in_format, out_pattern, out_format))
    nothing
end

"""
    run_sequence_file(settings_path)

Run a sequence settings document: the pipeline schema plus `Mode`,
`Input.Pattern`, `Input.Paths`, `Input.Times` and `Input.TimeFrom`. A document
using none of those behaves exactly as [`run_pipeline_file`](@ref).
"""
function run_sequence_file(settings_path::AbstractString)
    _check(ccall(_sym(:mio_sequence_pipeline_run_file), Cint, (Cstring,), settings_path))
    nothing
end

"""
    run_sequence_json(json_text)

[`run_sequence_file`](@ref) over JSON text.
"""
function run_sequence_json(json_text::AbstractString)
    _check(ccall(_sym(:mio_sequence_pipeline_run_json), Cint, (Cstring,), json_text))
    nothing
end
